### PR TITLE
Customer Facing Sitemap

### PR DESCRIPTION
**Pull request by teallarson**
## Change description

Adds a customer facing sitemap listing every page on www.

Docs are consolidated such that it shows the main landing pages that subpages are linked from.  I believe this is okay because they would be "resource pages" per Victorious' description.

Not sure whether to adjust the spacing of the columns -- they need to be this wide for the blog post titles, but the shorter items are somewhat awkwardly spaced.  Open to thoughts!

Screen Shot 2021-11-03 at 3 46 11 PM

### Development

- [x] Application changes have been tested appropriately

### Impact

- [x] Code follows company security practices and guidelines
- [x] Security impact of change has been considered
- [x] Performance impact of change has been considered
- [x] Possible migration needs considered (model migrations, config migrations, etc.)

Please explain any security, performance, migration, or other impacts if relevant:

### Code review

- [x] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached where applicable.
- [x] Relevant tags have been added to the PR (bug, enhancement, internal, etc.)

### CODE REVIEW

File: components/sitemap/sitemapBlogSection.tsx [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  entries: { [key: string]: BlogEntry[] };
+}) {
+  return (
+    <div className="homePageSection my-5">
+      <Container>
+        <div className="h2 col-10 mx-auto border-bottom">Blog</div>
+        <Col className="col-9 mx-auto">
+          <Row>
+            {Object.keys(entries).map((category) => {
+              return (
```
Comment (krishnaglick) [inline components/sitemap/sitemapBlogSection.tsx]: Minor nitpick: You can do away with the `{ return (` and just have the `(` portion

File: components/sitemap/sitemapBlogSection.tsx [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+                <Row className="col-12">
+                  <h3 className="col-12 mb-1 mt-3">{titleize(category)}</h3>
+                  {entries[category].map((entry) => {
+                    let path = `/blog/${entry.slug}`;
+                    return (
+                      <Col key={entry.slug} className="col-lg-6 col-12 my-1">
+                        <Link href={path}>{entry.title}</Link>
```
Comment (krishnaglick) [inline components/sitemap/sitemapBlogSection.tsx]:
```suggestion
                        <Link href={`/blog/${entry.slug}`}>{entry.title}</Link>
```
More inlining available here, no need to for `=> {`.
Can just have the jsx directly in the paren.

File: components/sitemap/sitemapBlogSection.tsx [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        <Col className="col-9 mx-auto">
+          <Row>
+            {Object.keys(entries).map((category) => {
+              return (
+                <Row className="col-12">
+                  <h3 className="col-12 mb-1 mt-3">{titleize(category)}</h3>
+                  {entries[category].map((entry) => {
+                    let path = `/blog/${entry.slug}`;
+                    return (
+                      <Col key={entry.slug} className="col-lg-6 col-12 my-1">
```
Comment (krishnaglick) [inline components/sitemap/sitemapBlogSection.tsx]: This needs a `key`. I think `entry.slug` is probably safe.

File: components/sitemap/sitemapBlogSection.tsx [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+}) {
+  return (
+    <div className="homePageSection my-5">
+      <Container>
+        <div className="h2 col-10 mx-auto border-bottom">Blog</div>
+        <Col className="col-9 mx-auto">
+          <Row>
+            {Object.keys(entries).map((category) => {
+              return (
+                <Row className="col-12">
```
Comment (krishnaglick) [inline components/sitemap/sitemapBlogSection.tsx]: This needs a `key`

File: components/sitemap/sitemapIntegrationsSection.tsx [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        <Col className="col-9 mx-auto">
+          <Row>
+            <h3 className="col-12 mb-1 mt-3">Sources</h3>
+            {integrations["sources"].map((integration) => {
+              let path = `/integrations/sources/${integration}`;
+              return (
+                <Col key={integration} className="col-md-6 col-12 my-1">
+                  <Link href={path}>{titleize(integration)}</Link>
+                </Col>
+              );
```
Comment (krishnaglick) [inline components/sitemap/sitemapIntegrationsSection.tsx]:
```
              let path = `/integrations/sources/${integration}`;
              return (
                <Col key={integration} className="col-md-6 col-12 my-1">
                  <Link href={path}>{titleize(integration)}</Link>
                </Col>
              );
```
Needs a `key`. And you can do inlining with that path value.

File: components/sitemap/sitemapIntegrationsSection.tsx [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+          </Row>
+          <Row>
+            <h3 className="col-12 mb-1 mt-3">Destinations</h3>
+            {integrations["destinations"].map((integration) => {
+              let path = `/integrations/destinations/${integration}`;
+              return (
+                <Col key={integration} className="col-md-6 col-12 my-1">
+                  <Link href={path}>{titleize(integration)}</Link>
+                </Col>
+              );
```
Comment (krishnaglick) [inline components/sitemap/sitemapIntegrationsSection.tsx]: Same here.

File: components/sitemap/sitemapIntegrationsSection.tsx [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+          </Row>
+          <Row>
+            <h3 className="col-12 mb-1 mt-3">Misc</h3>
+            {integrations["other"].map((integration) => {
+              let path = `/integrations/${integration}`;
+              return (
+                <Col key={integration} className="col-md-6 col-12 my-1">
+                  <Link href={path}>{titleize(integration)}</Link>
+                </Col>
+              );
```
Comment (krishnaglick) [inline components/sitemap/sitemapIntegrationsSection.tsx]: Same as above.

File: components/sitemap/sitemapSection.tsx [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        </div>
+        <Col className="col-9 mx-auto">
+          <Row>
+            {paths.map((pageName: string) => {
+              if (pageName.length === 0) return;
+              if (pageName === "trial_landing") return;
+              let path;
+              category === "other"
+                ? (path = `/${pageName}`)
+                : (path = `/${category}/${pageName}`);
```
Comment (krishnaglick) [inline components/sitemap/sitemapSection.tsx]:
```suggestion
              const path = category === "other"
                ? `/${pageName}`
                : `/${category}/${pageName}`;
```

File: components/sitemap/sitemapSection.tsx [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    <div style={style} className={classNames}>
+      <Container>
+        <div className="h2 col-10 mx-auto border-bottom">
+          {titleize(category)}
+        </div>
+        <Col className="col-9 mx-auto">
+          <Row>
+            {paths.map((pageName: string) => {
+              if (pageName.length === 0) return;
+              if (pageName === "trial_landing") return;
```
Comment (krishnaglick) [inline components/sitemap/sitemapSection.tsx]: You could combine these.

File: components/sitemap/sitemapSection.tsx [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+              let path;
+              category === "other"
+                ? (path = `/${pageName}`)
+                : (path = `/${category}/${pageName}`);
+              let pageTitle = titleize(pageName).replace(/-/g, " ");
```
Comment (krishnaglick) [inline components/sitemap/sitemapSection.tsx]:
```suggestion
              const pageTitle = titleize(pageName).replace(/-/g, " ");
```

File: utils/sitemap.ts [hunk 11]
```diff
@@ -81,3 +81,12 @@ export async function getSitemapStream() {
...
   return await streamToPromise(smStream).then((sm) => sm.toString());
 }
+
+export async function getPublicSitemap() {
+  const pages = await getPagePaths();
+  const integrations = await getStaticPaths();
+
+  const paths = pages.concat(integrations);
+
+  return paths;
```
Comment (krishnaglick) [inline utils/sitemap.ts]: Minor nitpick:
```suggestion
  return pages.concat(integrations);
```

File: utils/sitemap.ts [hunk 12]
```diff
@@ -27,7 +27,7 @@ export async function getBlogCategoryPaths(): Promise<PagePathsResponse> {
   return { paths, fallback: false };
 }
 
-async function getPagePaths() {
+export async function getPagePaths() {
```
Comment (krishnaglick) [inline utils/sitemap.ts:30]:
```suggestion
async function getPagePaths() {
```

File: utils/sitemap.ts [hunk 13]
```diff
@@ -43,7 +43,7 @@ async function getPagePaths() {
   return paths;
 }
 
-async function getStaticPaths() {
+export async function getStaticPaths() {
```
Comment (krishnaglick) [inline utils/sitemap.ts]:
```suggestion
async function getStaticPaths() {
```

File: components/sitemap/sitemapBlogSection.tsx [hunk 14]
```diff
@@ -0,0 +1,38 @@
+import { Container, Col, Row } from "react-bootstrap";
+import Link from "next/link";
+import { titleize } from "../../utils/inflectors";
+import { BlogEntry } from "../../utils/blogPosts";
+import ComparisonPage from "../solutions/comparisonPage";
+
+export default function SitemapBlogSection({
+  entries,
+}: {
+  entries: { [key: string]: BlogEntry[] };
```
Comment (krishnaglick) [inline components/sitemap/sitemapBlogSection.tsx]:
```suggestion
  entries: Record<string, BlogEntry[]>;
```

File: components/sitemap/sitemapBlogSection.tsx [hunk 15]
```diff
@@ -0,0 +1,38 @@
+import { Container, Col, Row } from "react-bootstrap";
+import Link from "next/link";
+import { titleize } from "../../utils/inflectors";
+import { BlogEntry } from "../../utils/blogPosts";
+import ComparisonPage from "../solutions/comparisonPage";
```
Comment (krishnaglick) [inline components/sitemap/sitemapBlogSection.tsx]: This is unused

File: pages/public-sitemap.tsx [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+  );
+}
+
+export async function getStaticProps(context) {
+  const sitemapPath = path.join(process.cwd(), "public", "public-sitemap.json");
+
+  const data = await fs.readFileSync(sitemapPath, "utf8");
+  let paths = JSON.parse(data);
+
+  let ret = {};
```
Comment (krishnaglick) [inline pages/public-sitemap.tsx]: This is pretty confusing. I'd be happy to pair to see if we can simplify it!

File: utils/sitemap.ts (hunk 12 above)
Comment (krishnaglick) [reply utils/sitemap.ts:30]: I think we were trying something when I added this, but doesn't look like it's needed any more!

**Review by andyjih — CHANGES_REQUESTED** [review]
Overall it looks good! A few things:
1. I'd like not to include any terms of service that are tied to a specific date. I'm keeping those mostly as records for specific contracts, but I don't want to have that indexed for SEO purposes.



2. It seems like some Integrations pages and Blogposts are missing.

**Review by andyjih — APPROVED** [review]
Oh, nevermind, looks like a commit just fixed it. lgtm!

**Comment by teallarson** [comment]
> Oh, nevermind, looks like a commit just fixed it. lgtm!

Yeah. KC and I were cleaning a few things up!  Removed anything with `/terms-*` so future versioned docs should be good.  All the docs and blog posts are there now again :)

File: __tests__/utils/inflectors.ts [hunk 17]
```diff
@@ -9,36 +13,37 @@ describe("#titleize", () => {
     const res = titleize("_hello__world!_");
     expect(res).toBe("Hello World!");
   });
-  test("does not remove hyphens", () => {
+  test("removes hyphens", () => {
     const res = titleize("hello-world");
-    expect(res).toBe("Hello-World");
+    expect(res).toBe("Hello World");
```
Comment (krishnaglick) [inline __tests__/utils/inflectors.ts:18]: We should probably add a test for the new `breadcrumbTitleize`

File: pages/public-sitemap.tsx [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+  docs: Record<string, SitemapItem[]>;
+  legal: SitemapItem[];
+  solutions: SitemapItem[];
+  other: SitemapItem[];
+  integrations: {
+    sources: SitemapItem[];
+    destinations: SitemapItem[];
+    guides: SitemapItem[];
+  };
+  blog: Record<keyof typeof badgeTypes, SitemapItem[]>;
```
Comment (krishnaglick) [inline pages/public-sitemap.tsx:25]: The intent here is that blogs should be limited to our existing categories, docs should be dynamic to accommodate us adding or removing things as necessary in the future (think Profiles -> Records), and integrations should be fairly static in their categories.

File: pages/public-sitemap.tsx [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+          name,
+        });
+      }
+    } else {
+      throw new Error("Empty path: " + path);
+    }
+  });
+
+  blogPosts.forEach((blogPost) => {
+    const [category] = blogPost.tags;
```
Comment (krishnaglick) [inline pages/public-sitemap.tsx:141]: Use the first tag for a post as its category.

File: scripts/generate_static_files.ts [hunk 20]
```diff
@@ -3,11 +3,12 @@ import path from "path";
...
 import { getFeed as whatsNewFeed } from "../utils/releaseNotes";
 import { getFeed as blogFeed } from "../utils/blogPosts";
-import { getSitemapStream } from "../utils/sitemap";
+import { getPublicSitemap, getSitemapStream } from "../utils/sitemap";
 
 async function build() {
   const feeds = writePublic(await feedContent());
   const sitemap = writePublic(await sitemapContent());
+  const publicSitemap = writePublic(await jsonSitemap());
   return Promise.all([feeds]);
```
Comment (krishnaglick) [inline scripts/generate_static_files.ts:12]: It's unclear to me why we have `sitemap` here, but not included in the `Promise.all` @evantahler

File: utils/inflectors.ts [hunk 21]
```diff
@@ -1,4 +1,14 @@
...
+  return str
+    .replace(/^[\s_]+|[\s_]+$/g, "")
+    .replace(/[_\s]+/g, " ")
+    .replace(/-/g, " ")
+    .replace(/^[\w]|[ _-]\w/g, (m) => {
+      return m.toUpperCase();
+    });
+}
+
+export function breadcrumbTitleize(str: string) {
```
Comment (krishnaglick) [inline utils/inflectors.ts]: We could probably make titleize call breadcrumbTitleize, and include the `.replace(/-/g, " ")` portion.
More DRY.

File: utils/mdxUtils.ts [hunk 22]
```diff
@@ -10,7 +10,7 @@ import rehypeAutolinkHeadings from "rehype-autolink-headings";
 import rehypeSlug from "rehype-slug";
 import externalLinks from "remark-external-links";
 
-import { titleize } from "../utils/inflectors";
+import { titleize, breadcrumbTitleize } from "../utils/inflectors";
```
Comment (krishnaglick) [inline utils/mdxUtils.ts]: Doesn't look like `breadcrumbTitleize` is used in here.

File: __tests__/utils/inflectors.ts (hunk 17 above)
Comment (teallarson) [reply __tests__/utils/inflectors.ts:18]: I took it back out.  The place I thought was leaving hyphens in was actually removing them.  Meaning every place in the code base was removing hyphens after calling this function.

File: utils/inflectors.ts (hunk 21 above)
Comment (teallarson) [reply utils/inflectors.ts]: removed this function... the only places using it were also calling .replace() to remove hyphens immediately

**Review by krishnaglick — APPROVED** [review]

**Pull request merged**